src/v/cloud_storage/materialized_resources.cc

```diff
@@ -379,4 +399,47 @@ void materialized_resources::maybe_trim_segment(
     }
 }
 
+class throttled_dl_source : public ss::data_source_impl {
+public:
+    explicit throttled_dl_source(ss::input_stream<char> st, token_bucket<>& tb)
+      : _src(std::move(st))
+      , _tokens(tb) {}
+
+    ss::future<ss::temporary_buffer<char>> get() override {
+        return ss::with_gate(_gate, [this] {
+            return _src.read().then([this](ss::temporary_buffer<char> buf) {
+                if (!_tokens.try_throttle(buf.size())) {
+                    // TODO: throw different exception
```

Comment:
nit: +1 maybe we can make use of a new case of cloud_storage::download_exception or somesuch? It'll be useful to catch these at a higher level and pass the throttle on to clients

Reply:
reworked this